src/gameState.ts

```diff
@@ -0,0 +1,299 @@
+                    for (let i:number = 0; i<8; i++){
+                        let dr:number = row + x[i];
+                        let dc:number = column + y[i];
+                        if (this.isValidTarget(new Position({"Row":GameState.numberToLetter[dr], "Column":dc+1}))){
+                            this.board[dr][dc] = -1;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public sink(startRow:number, startColumn:number, endRow:number, endColumn:number):void{
+        for (let row:number=startRow; row<=endRow; row++){
+            for (let column:number = startColumn; column <= endColumn; column++){
+                this.board[row][column] = 2;
+            }
+        }
+    }
+
+    public isSunken(startRow:number, startColumn:number, endRow:number, endColumn:number):boolean{
+        if ((startRow==endRow) && (startColumn == endColumn)){
+            return false;
+        }
+        if (startRow!=endRow){
+            if (Math.abs(endRow - startRow)+1 == this.remainingShips[this.remainingShips.length-1]){
+                return true;
+            }
+            if ((startRow == 0) || (this.board[startRow-1][startColumn]==-1)){
+                if ((endRow == 9) || (this.board[endRow+1][startColumn]==-1)){
+                    return true;
+                }
+            }
+        }
+        else {
+            if (Math.abs(endColumn - startColumn)+1 == this.remainingShips[this.remainingShips.length-1]){
+                return true;
+            }
+            if ((startColumn == 0) || (this.board[startRow][startColumn-1]==-1)){
+                if ((endColumn == 9) || (this.board[startRow][endColumn+1]==-1)){
+                    return true;
+                }
+            }
+        }
+        return false
+    }
+
+    // returns a position on the board that is a 1 i.e. a ship we hit but haven't sunk yet (if it exists)
+    public getHitPosition():Position{
+        for (let row:number = 0; row<10; row++){
+            for (let column:number = 0; column<10; column ++){
+                if (this.board[row][column] == 1){
+                    return new Position({"Row":GameState.numberToLetter[row], "Column":column+1});
+                }
+            }
+        }
+        return null
+    }
+
+    // given hitPosition, returns a position with a hit that is a neighbour of hitPosition (if such position exists)
+    public getNeighbourHitPosition(hitPosition:Position):Position{
+        let row:number = hitPosition.row;
+        let column:number = hitPosition.column;
+        if (column<9){
+            if (this.board[row][column+1]==1){
+                return new Position({"Row":GameState.numberToLetter[row], "Column":column+2});
+            }
+        }
+        if (row<9){
+            if (this.board[row+1][column]==1){
+                return new Position({"Row":GameState.numberToLetter[row+1], "Column":column+1});
+            }
+        }
+        return null;
```

Comment:
Returning null isn't great - it can cause all kinds of issues!

Could you possibly check whether you're able to get a position before returning one here?